akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -283,14 +291,21 @@ private[akka] object StreamLayout {
 
     override def toString =
       s"""
-        | Module: ${this.attributes.nameOption.getOrElse("unnamed")}
-        | Modules: ${subModules.toSeq.map(m ⇒ "   " + m.attributes.nameOption.getOrElse(m.getClass.getName)).mkString("\n")}
+        | Module: ${this.attributes.nameOrDefault("unnamed")}
+        | Modules: ${subModules.iterator.map(m ⇒ "   " + m.attributes.nameOrDefault(m.getClass.getName)).mkString("\n")}
         | Downstreams:
-        | ${downstreams.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
+        | ${downstreams.iterator.map { case (in, out) ⇒ s"   $in -> $out" }.mkString("\n")}
         | Upstreams:
-        | ${upstreams.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
+        | ${upstreams.iterator.map { case (out, in) ⇒ s"   $out -> $in" }.mkString("\n")}
       """.stripMargin
 
+    override def describe: String =
+      attributes.nameOrDefault(null) match {
+        case null ⇒ "(" + subModules.iterator.map(_.describe).mkString(" ~> ") + ")"
```

Comment:
So, how I did this before: there was a LinearComposite which was a CompositeModule variant that was always linear, only changing to CompositeModule if linearity was violated (i.e. general graph). The toString on LinearComposite then was easy. I sadly don't find the branch, I hope I didn't delete it. I will need a LinearComposite probably anyway to help fusing.

This won't help with nonlinear modules of course, that belongs to the nasty world of graph layout algorithms. I had some ideas how to make something useful for debug that looks reasonable for most of the cases without going too far with graph layout algos. Instead we will provide a DOT printer that you can use with Graphviz or some other layout tool.